test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java

```diff
@@ -2622,4 +2631,72 @@ protected ClusterState getClusterState() {
+        try {
+            assertBusy(() -> {
+                final ClusterState clusterState = client(internalCluster().getClusterManagerName()).admin()
+                    .cluster()
+                    .prepareState()
+                    .get()
+                    .getState();
+                for (IndexRoutingTable indexRoutingTable : clusterState.routingTable()) {
+                    for (IndexShardRoutingTable shardRoutingTable : indexRoutingTable) {
+                        final ShardRouting primaryRouting = shardRoutingTable.primaryShard();
+                        if (primaryRouting.state().toString().equals("STARTED")) {
+                            final String indexName = primaryRouting.getIndexName();
+                            if (isSegmentReplicationEnabledForIndex(indexName)) {
+                                final List<ShardRouting> replicaRouting = shardRoutingTable.replicaShards();
+                                final IndexShard primaryShard = getIndexShard(clusterState, primaryRouting, indexName);
+                                final int primaryDocCount = getDocCountFromShard(primaryShard);
+                                for (ShardRouting replica : replicaRouting) {
+                                    if (replica.state().toString().equals("STARTED")) {
+                                        IndexShard replicaShard = getIndexShard(clusterState, replica, indexName);
+                                        final int replicaDocCount = getDocCountFromShard(replicaShard);
+                                        assertEquals("Replica doc count doesn't match primary", primaryDocCount, replicaDocCount);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }, 30, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Checks if Segment Replication is enabled on Index.
+     */
+    protected boolean isSegmentReplicationEnabledForIndex(String index) {
+        return clusterService().state().getMetadata().isSegmentReplicationEnabled(index);
+    }
+
+    protected IndexShard getIndexShard(ClusterState state, ShardRouting routing, String indexName) {
+        return getIndexShard(state.nodes().get(routing.currentNodeId()).getName(), routing.shardId(), indexName);
+    }
+
+    /**
+     * Fetch IndexShard by shardId, multiple shards per node allowed.
+     */
+    protected IndexShard getIndexShard(String node, ShardId shardId, String indexName) {
+        final Index index = resolveIndex(indexName);
+        IndicesService indicesService = internalCluster().getInstance(IndicesService.class, node);
+        IndexService indexService = indicesService.indexServiceSafe(index);
+        final Optional<Integer> id = indexService.shardIds().stream().filter(sid -> sid == shardId.id()).findFirst();
```

Comment:
Use `.equals` for non-primitive types (I think your code works because Java caches small integer values and always uses the same instances, but this isn't guaranteed to work for larger values).

```suggestion
        final Optional<Integer> id = indexService.shardIds().stream().filter(sid -> sid.equals(shardId.id())).findFirst();
```

Reply:
Done